feat: tidy up `ToyExecutionEnvironment` and remove `ToyWorld`

DavePearce opened the pull request:
This tidies up the `ToyExecutionEnvironment` in several ways.  Firstly, it updates the manner in which the `MainnetTransactionProcessor` is created by reusing as much from BESU as possible; secondly, it removes `ToyWorld` altogether (as this is no longer needed); finally, it removes `ToyExecutionEnvironment.execute()` since this is no longer being used.

DavePearce committed:
Tidy up `ToyExecutionEnvironment`

This tidies up the `ToyExecutionEnvironment` in several ways.  Firstly,
it updates the manner in which the `MainnetTransactionProcessor` is
created by reusing as much from BESU as possible; secondly, it removes
`ToyWorld` altogether (as this is no longer needed); finally, it removes
`ToyExecutionEnvironment.execute()` since this is no longer being used.

DavePearce committed:
Support Mainnet vs Sepolia Tests

This adds support for distinguishing between Mainnet and Sepolia tests.
This is helpful for the odds cases where we want a replay which was
inteded for sepolia.

DavePearce committed:
Remove `getMainnetTransactionProcessorOrig()`

This removes the now redundant (original) method for getting the mainnet
transaction processor.

DavePearce committed:
Apply spotless

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/blockcapture/snapshots/ConflationSnapshot.java:
```diff
@@ -29,4 +34,44 @@ public record ConflationSnapshot(
+   */
+  public BlockHashLookup toBlockHashLookup() {
+    BlockHashMap map = new BlockHashMap();
+    // Initialise block hashes.  This can be null for replays which pre-date support for block hash
```
It may be worth re-capturing them. We have lists and @gusiri 's `bulk_capture.sh` script.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/blockcapture/snapshots/ConflationSnapshot.java:
```diff
@@ -29,4 +34,44 @@ public record ConflationSnapshot(
+  public BlockHashLookup toBlockHashLookup() {
+    BlockHashMap map = new BlockHashMap();
+    // Initialise block hashes.  This can be null for replays which pre-date support for block hash
+    // capture and, hence, we must support this case (at least for now).
```
It may be worth re-capturing them. We have lists and @gusiri 's `bulk_capture.sh` script.

DavePearce replied:
Yeah, I agree with that.  But, we need to retain the ability to run without for now, otherwise it will break stuff for others.

OlivierBBB commented on testing/src/main/java/net/consensys/linea/testing/ReplayExecutionEnvironment.java:
```diff
@@ -0,0 +1,286 @@
+   * even if the tx outcome differs from what actually occurred on the relevant chain (e.g.
+   * mainnet). Such scenarios have arisen, for example, when <code>ZkTracer</code> has an unexpected
+   * side-effect on tx execution (which it never should).
+   */
```
I guess we don't know what could cause any remaining side effects at this point do we ? It may be interesting to have more information than just a boolean.

DavePearce replied:
We don't know what we don't know!  But, also, there are other scenarios you should consider:

1.  Someone from the team could accidentally reintroduce a call to a side-effecting method (such as `gasAvailableForChildCall()`).  
2. BESU itself could be changed in a way that makes a previously pure method now have side-effects.  

We want to get notified of this immediately, in the event anything like this happens.

OlivierBBB commented on testing/src/main/java/net/consensys/linea/testing/ReplayExecutionEnvironment.java:
```diff
@@ -0,0 +1,286 @@
+    this.checkTracer();
+  }
+
+  public void replay(BigInteger chainId, final Reader replayFile, String inputFilePath) {
```
These two methods should be mergeable. I don't know if this is a use case for `Optional<String>` or for `String... inputFilePath` but there ought to be a way to unite these.

DavePearce replied:
So, yeah, I do agree with this.  I looked into it briefly, but I need to investigate in more detail what exactly is going on with the `inputFilePath`.  For now, I didn't think it was worth investing time in.

OlivierBBB commented on testing/src/main/java/net/consensys/linea/testing/ReplayExecutionEnvironment.java:
```diff
@@ -0,0 +1,286 @@
+   * @param world The world to be initialised.
+   * @param conflation The conflation from which to initialise.
+   */
+  private static void initWorld(WorldUpdater world, final ConflationSnapshot conflation) {
```
Does a world that was thusly initialized know about precompiles ? I ask because in our old setup we had to invoke a special function to make the execution environment and the world aware of precompiles.

OlivierBBB reviewed (approved):
LGTM (however: I mostly skipped reading `getMainnetTransactionProcessor`)

DavePearce committed:
Fix broken replay test.

`modexpTriggeringNonAlignedFirstLimbSingleSourceMmuModexp` was missing
the chain identifier.  I'm assuming LINEA_MAINNET for now.

OlivierBBB committed:
Merge branch 'arith-dev' into feat/issue-1048/remove-toyworld

gusiri commented on arithmetization/src/test/java/net/consensys/linea/zktracer/ReplayTests.java:
```diff
@@ -200,54 +206,54 @@ void fatMxp() {
   @Test
   void bulkReplay() {
     // bulkReplay("./src/test/resources/replays");
```
@DavePearce 
Can we also update this comment to inlcude LINEA_MAINNET?
// bulkReplay(LINEA_MAINNET, "./src/test/resources/replays");

DavePearce replied:
Yes, we can do.  Create a separate issue and I will add support there.  I haven't completely grokked the `builkReplay` system, so I need to talk it over with you first, etc.

gusiri replied:
Thank you 👍

DavePearce replied:
It would be good also to merge the two replay methods somehow.  I will have a look at that also
